proxy.go

```diff
@@ -188,6 +216,207 @@ func (proxy *proxy) releaseToken(token Token) (*tokenInfo, error) {
+func (proxy *proxy) restoreState() bool {
+	proxyStateFilePath := storeStateDir + "proxy_state.json"
+	if _, err := os.Stat(storeStateDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(storeStateDir, 0755); err != nil {
+			proxyLog.Errorf("Couldn't create directory %s: %v",
+				storeStateDir, err)
+		}
+		return false
+	}
+	if _, err := os.Stat(proxyStateFilePath); os.IsNotExist(err) {
+		return false
+	}
+
+	fdata, err := ioutil.ReadFile(proxyStateFilePath)
+	if err != nil {
+		proxyLog.Errorf("Couldn't recover from %s: %v", proxyStateFilePath, err)
+		return false
+	}
+
+	var proxyState proxyStateOnDisk
+	err = json.Unmarshal(fdata, &proxyState)
+	if err != nil {
+		proxyLog.Errorf("Couldn't unmarshal %s: %v", proxyStateFilePath, err)
+		return false
+	}
+	proxyLog.Debugf("proxy: %+v", proxyState)
+
+	if len(proxyState.ContainerIDs) == 0 {
+		return false
+	}
+	proxyLog.Warn("Recovering proxy state from: ", proxyStateFilePath)
+	proxy.socketPath = proxyState.SocketPath
+	proxy.enableVMConsole = proxyState.EnableVMConsole
+
+	for _, contID := range proxyState.ContainerIDs {
+		go restoreVMState(proxy, contID)
+	}
+
+	return true
+}
+
+func (proxy *proxy) storeState() {
+	proxyStateFilePath := storeStateDir + "proxy_state.json"
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	// if there are 0 VMs then remove state from disk
+	if (len(proxy.vms)) == 0 {
+		if err := os.Remove(proxyStateFilePath); err != nil {
+			proxyLog.Errorf("Can not remove %s: %v", proxyStateFilePath, err)
+		}
+		return
+	}
+
+	proxyState := &proxyStateOnDisk{
+		SocketPath:      proxy.socketPath,
+		EnableVMConsole: proxy.enableVMConsole,
+		ContainerIDs:    make([]string, 0, len(proxy.vms)),
+	}
+	for cid := range proxy.vms {
+		proxyState.ContainerIDs = append(proxyState.ContainerIDs, cid)
+	}
+
+	data, err := json.MarshalIndent(proxyState, "", "\t")
+	if err != nil {
+		proxyLog.Errorf("Couldn't marshal proxy state %+v", proxyState)
+	}
+	if err = ioutil.WriteFile(proxyStateFilePath, data, 0644); err != nil {
+		proxyLog.Errorf("Couldn't store proxy state to %s: %v",
+			proxyStateFilePath, err)
+	}
+}
+
+// Represents vm struct on disk
+type vmStateOnDisk struct {
+	RegisterVM api.RegisterVM `json:"registerVM"`
+	Tokens     []string       `json:"tokens"`
+}
+
+func vmStateFilePath(id string) string {
+	return storeStateDir + "vm_" + id + ".json"
+}
+
+func storeVMState(vm *vm, tokens []string) {
+	odVM := vmStateOnDisk{
+		RegisterVM: api.RegisterVM{
+			ContainerID: vm.containerID,
+			CtlSerial:   vm.hyperHandler.GetCtlSockPath(),
+			IoSerial:    vm.hyperHandler.GetIoSockPath(),
+			Console:     vm.console.socketPath,
+		},
+		Tokens: tokens,
+	}
+	o, err := json.MarshalIndent(&odVM, "", "\t")
+	if err != nil {
+		proxyLog.WithField("vm", vm.containerID).Warn("Couldn't marshal VM state")
+	}
+	storeFile := vmStateFilePath(vm.containerID)
+	if err = ioutil.WriteFile(storeFile, o, 0644); err != nil {
```

Comment:
Another perm which should probably be `0640`.

Reply:
Agree, will do.